source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -35,6 +35,19 @@ const Common::Redis::RespValue& getRequest(const RespVariant& request) {
 }
 
 static uint16_t default_port = 6379;
+
+bool isClusterProvidedLb(const Upstream::ClusterInfo& info) {
+  const auto lb_type = info.lbType();
+  bool cluster_provided_lb = lb_type == Upstream::LoadBalancerType::ClusterProvided;
+  if (lb_type == Upstream::LoadBalancerType::LoadBalancingPolicyConfig) {
+    auto* typed_lb_factory = info.loadBalancerFactory();
+    RELEASE_ASSERT(typed_lb_factory != nullptr, "ClusterInfo should contain a valid factory");
```

Comment:
if a requested factory is compiled out shouldn't we gracefully fail (throw/status fail) rather than crash?

Reply:
Reasonable suggestion. But if there is invalid lb factory that is compiled out, the cluster loading will failed when we create the cluster info.
So, this ASSERT should never happen and this why I used RELEASE_ASSERT here. If it's triggered, than it a bug rather than compile issue.

Reply:
My concern is if the assertion ever fails (the utility is refactored) we'll have a production crash where we're trying really hard to avoid having release assert. At least let's add a comment what should happen here.  best case it'd be nice to regression test having a redis e2e test which tries to create a non-linked cluster to future-proof against harmful refactors.

Reply:
got it. I can replace the RELEASE_ASSERT with a `ENVOY_BUG` and add additional check in the constructor of `ClusterInfoImpl`.

But it's hard to add a test case to test it. As I said in previous comments, the constructor of `ClusterInfoImpl` will throw if there is no valid factory and the creation of cluster will fail. The `isClusterProvidedLb()` will never be called for a fail-created cluster.

Reply:
In fact, we do similiar `RELEASE_ASSERT` in other positions. See

https://github.com/envoyproxy/envoy/blob/b31401c94476bc52743f25021c50de8490abe63b/source/common/upstream/cluster_manager_impl.cc#L948-L958